avoid database error when recording task exception

jimabramson opened the pull request:
@nasthagiri @dan-f  please review (non-urgent, but very small change)

this should fix a problem where processing a course rerun throws an exception, but the attempt to store that outcome in the database fails because the exception message is too long.

nasthagiri commented on cms/djangoapps/contentstore/tasks.py:
```diff
@@ -58,7 +58,7 @@ def rerun_course(source_course_key_string, destination_course_key_string, user_i
             pass
 
-        return "exception: " + unicode(exc)
+        return ("exception: " + unicode(exc))[:999]  # don't exceed 1000 chars
```
Hmm.. these return statements are used only by tests currently.  Is it really this line that is causing the issue?

The data that is stored in the database is actually on L51 above (see implementation of `CourseRerunState.objects.failed`).  What is the stack trace that you got with the database error?

dan-f replied:
Why the 1000 character limit? Is that a SQL thing? Could you use a constant? Maybe it could live in `common/djangoapps/util/db.py`

jimabramson replied:
@nasthagiri hm, thanks for the catch - I think I'm doing the right thing but in the wrong place.  

Here's a gist of the trace: https://gist.github.com/jimabramson/fa4839466ec04d5b7786  ...which suggests the string truncation actually belongs in course_action_state/managers.py.

jimabramson replied:
@dan-f yes, SQL thing, see the above gist.

TBH this is kind of a band-aid fix in my view.  If we intend to store entire tracebacks in a database column that column ought to be provisioned with enough space to accomodate one, otherwise we should log it to a file where it's going to be more reasonably accessible via better tools for the job (i.e. splunk).

my main concern with this PR is that we might have failures creating reruns and lose any info we could use to diagnose those failures.

jimabramson commented:
@nasthagiri @dan-f  Thanks for taking the look - unless there's an alternative quick fix recommended I will probably close this PR and file a ticket instead as I don't have a lot of time to work deeply on this, and I doubt either of you do either :-)

nasthagiri commented:
@jimabramson The alternative quick fix would be to apply the same fix but at this location:
https://github.com/edx/edx-platform/blob/master/common/djangoapps/course_action_state/managers.py#L146

jimabramson commented:
@nasthagiri @dan-f i went back and changed the fix, and added a test.  please review?

NOTE - wasn't sure whether it made more sense to merge the new test into the previous one, but opted not to, since i think it might make for a difficult read to test all these things in one method.

nasthagiri commented on cms/djangoapps/contentstore/tests/test_contentstore.py:
```diff
@@ -1721,6 +1721,26 @@ def test_rerun_error(self):
 
+    def test_rerun_error_trunc_message(self):
+        """
+        CourseActionUIState.message only accomodates 1000 characters, but is sometimes
```
Nit: s/accomodates/accommodates

nasthagiri commented on common/djangoapps/course_action_state/managers.py:
```diff
@@ -143,7 +143,7 @@ def failed(self, course_key):
             course_key=course_key,
             new_state=self.State.FAILED,
-            message=traceback.format_exc(),
+            message=traceback.format_exc()[-1000:],  # truncate to fit CHAR(1000)
```
Can we declare a constant for this number in this file and then use that constant in both the test and here?

dan-f replied:
+1 on @nasthagiri's comment.

nasthagiri commented:
:+1: Two nits, otherwise looks good.

dan-f commented:
:+1:

jimabramson commented:
thanks for review.  updated with changes.

nasthagiri commented:
:+1:

jimabramson commented:
updated to fix quality complaint.
